cpu/atmega_common/periph/rtt.c

```diff
@@ -0,0 +1,337 @@
+    rtt_state.ext_comp[1] = (uint8_t)(alarm >> 16);
+    rtt_state.ext_comp[0] = (uint8_t)(alarm >> 8);
+    OCR2A = (uint8_t)alarm;
+
+    /* Wait until alarm value takes effect */
+    DEBUG("RTT sleeps until safe power-save\n");
+    while( ASSR & (1 << OCR2AUB) ) {
+#ifdef MODULE_XTIMER
+        /* Sleep for 2 RTC cycles */
+        xtimer_usleep(70 * US_PER_MS);
+#else
+        thread_yield();
+#endif
+    }
+
+    /* Enable alarm interrupt only if it will trigger before overflow */
+    if ( rtt_state.ext_comp[2] == rtt_state.ext_cnt[2]
+      && rtt_state.ext_comp[1] == rtt_state.ext_cnt[1]
+      && rtt_state.ext_comp[0] == rtt_state.ext_cnt[0] ) {
+        TIMSK2 |= (1 << OCIE2A);
+    }
+}
+
+uint32_t rtt_get_alarm(void) {
+    return (((uint32_t)rtt_state.ext_comp[2] << 24)
+          | ((uint32_t)rtt_state.ext_comp[1] << 16)
+          | ((uint32_t)rtt_state.ext_comp[0] << 8)
+          |  (uint32_t)OCR2A);
+}
+
+void rtt_clear_alarm(void) {
+    /* Disable alarm interrupt */
+    TIMSK2 &= ~(1 << OCIE2A);
+
+    rtt_state.alarm_cb = NULL;
+    rtt_state.alarm_arg = NULL;
+}
+
+void rtt_poweron(void) {
+    power_timer2_enable();
+}
+
+void rtt_poweroff(void) {
+    power_timer2_disable();
+}
+
+ISR(TIMER2_OVF_vect) {
+    __enter_isr();
+    /* Enable RTT alarm if overflowed enough times */
+    if ((rtt_state.ext_comp[2] == rtt_state.ext_cnt[2])
+     && (rtt_state.ext_comp[1] == rtt_state.ext_cnt[1])
+     && (rtt_state.ext_comp[0] == rtt_state.ext_cnt[0])) {
+        TIMSK2 |= (1 << OCIE2A);
+    }
+
+#if RTC_NUMOF
+    /* Increment RTC by 8 seconds */
+    atmega_rtc_incr();
+#endif
+
+    /* Virtual 32-bit timer overflow */
+    if (rtt_state.ext_cnt[0] == 0xff) {
+        rtt_state.ext_cnt[0] = 0;
+
+        if (rtt_state.ext_cnt[1] == 0xff) {
+            rtt_state.ext_cnt[1] = 0;
+
+            if (rtt_state.ext_cnt[2] == 0xff) {
+                rtt_state.ext_cnt[2] = 0;
+
+                /* Execute callback */
+                if (rtt_state.overflow_cb != NULL) {
+                    rtt_state.overflow_cb(rtt_state.overflow_arg);
+                }
+            } else {
+                rtt_state.ext_cnt[2]++;
+            }
+        } else {
+            rtt_state.ext_cnt[1]++;
+        }
+    } else {
+        rtt_state.ext_cnt[0]++;
+    }
+
+    /* Wait until it is safe to re-enter power-save */
+    TCCR2A = TCCR2A;
+    while( ASSR & (1 << TCR2AUB) ) {
+        if (sched_context_switch_request) {
+            thread_yield();
+        }
+    }
+    __exit_isr();
+}
+
+ISR(TIMER2_COMPA_vect) {
+    __enter_isr();
+    /* Disable alarm until overflowed enough times */
+    TIMSK2 &= ~(1 << OCIE2A);
+
+    /* Execute callback */
+    if (rtt_state.alarm_cb != NULL) {
+        rtt_state.alarm_cb(rtt_state.alarm_arg);
+    }
+
+    /* Wait until it is safe to re-enter power-save */
+    TCCR2A = TCCR2A;
+    while( ASSR & (1 << TCR2AUB) ) {
+        if (sched_context_switch_request) {
```

Comment:
why is this in the loop?

Reply:
According to the datasheet: if the MCU is woken from sleep by the asynchronous timer (as we have here in the interrupt), and then sleeps again before 1 cycle of the timer has passed, it doesn't reset the interrupt logic and will fail to wake a second time when an interrupt should have been generated. I will add some comments to make the purpose of each wait clearer.

Reply:
Ok, so ```while( ASSR & (1 << TCR2AUB) ) {}``` does the waiting. But why is the ```if (sched_... ``` in that while loop? It could be right after? Isn't ```thread_yield()``` executed multiple times?

Reply:
Oh. That was just because I saw that it was done that way in other interrupt routines. I assumed that the effect would be to switch threads; the switching would unset the context request; and then it would spin without switching threads until the check passed. I would use xtimer, but the docs for it say not to use it in interrupts because it will just spin anyway.

Reply:
Removed.